nexus/src/app/sagas/instance_create.rs

```diff
@@ -1144,17 +1178,136 @@ mod test {
                 .await
                 .expect_err("Saga should have failed");
 
-            let datastore = nexus.datastore();
+            verify_clean_clate(&cptestctx).await;
+        }
+    }
+
+    #[nexus_test(server = crate::Server)]
+    async fn test_action_failure_can_unwind_idempotently(
+        cptestctx: &ControlPlaneTestContext,
+    ) {
+        DiskTest::new(cptestctx).await;
+        let log = &cptestctx.logctx.log;
 
-            // Check that no partial artifacts of instance creation exist
-            assert!(no_instance_records_exist(datastore).await);
-            assert!(no_network_interface_records_exist(datastore).await);
-            assert!(no_external_ip_records_exist(datastore).await);
-            assert!(disk_is_detached(datastore).await);
-            assert!(
-                no_instances_or_disks_on_sled(&cptestctx.sled_agent.sled_agent)
-                    .await
+        let client = &cptestctx.external_client;
+        let nexus = &cptestctx.server.apictx.nexus;
+        let project_id = create_org_project_and_disk(&client).await;
+
+        // Build the saga DAG with the provided test parameters
+        let opctx = test_opctx(&cptestctx);
+
+        let params = new_test_params(&opctx, project_id);
+        let dag = create_saga_dag::<SagaInstanceCreate>(params).unwrap();
+
+        // The "undo_node" should always be immediately preceding the
```

Comment:
While perfectly reasonable with our current implementation, it makes me a tad uncomfortable that we are expecting a steno internal Dag structure to look a certain way for this test.

I'm not really sure the best way to handle this, except for perhaps first classing this type of injection in steno. I"m curious to hear @davepacheco's thoughts when he gets back, although I don't think we should hold up the PR for this.

Reply:
We aren't expecting anything of the internal DAG structure, other than "it's ordered". "Which node is which" is entirely up to the test -- the *test* is making the call "whenever we fail, the node preceding the failure node should have an undo action called twice to test idempotency".

We can arbitrarily change the saga DAG, and this test should be able to remain unchanged.

All nodes we're iterating over are either actions, constants, or subsagas, as discussed in https://github.com/oxidecomputer/steno/pull/67 .

> I'm not really sure the best way to handle this, except for perhaps first classing this type of injection in steno.

What would this mean to you? I thought the APIs I added to steno to [iterate over the dag](https://github.com/oxidecomputer/steno/commit/666141615d18dde53f0d00bc4c4f9b474af006e7), [expose indices which could be targets for injection](https://github.com/oxidecomputer/steno/commit/4cebb996217453b79896b53c9c2026007f2e69e8), and add the ability to [inject repetitions](https://github.com/oxidecomputer/steno/commit/53129dd9c011d1f9d2fd781676243198e1e297bc) alongside the already-existing "error injection" counted as first-class support?

Reply:
> We aren't expecting anything of the internal DAG structure, other than "it's ordered". "Which node is which" is entirely up to the test -- the test is making the call "whenever we fail, the node preceding the failure node should have an undo action called twice to test idempotency".

> We can arbitrarily change the saga DAG, and this test should be able to remain unchanged.

Yup, thinking about this more, you are 100% correct.

> What would this mean to you? I thought the APIs I added to steno to [iterate over the dag](https://github.com/oxidecomputer/steno/commit/666141615d18dde53f0d00bc4c4f9b474af006e7), [expose indices which could be targets for injection](https://github.com/oxidecomputer/steno/commit/4cebb996217453b79896b53c9c2026007f2e69e8), and add the ability to [inject repetitions](https://github.com/oxidecomputer/steno/commit/53129dd9c011d1f9d2fd781676243198e1e297bc) alongside the already-existing "error injection" counted as first-class support?

I was overthinking this. What you did makes sense. My initial thought was something unnecessarily complicated. We'd inject errors then iterate again, reading the annotated error nodes and getting the prior nodes affected by the errors. However, as you point out that's unnecessary, since we know based on the dag that the node immediately prior to the error node will be undone.

Sorry for unnecessarily getting your hackles up.

Reply:
No worries! Just want to make sure I wasn't missing a cleaner solution - probing at what's going on is always appreciated